pkg/macaroon/types.proto

```diff
@@ -0,0 +1,41 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+syntax = "proto3";
+
+package macaroon;
+
+import "gogo.proto";
+import "google/protobuf/timestamp.proto";
+
+message Caveat {
+  // if any of these three are set, disallow that type of access
+  bool disallow_reads = 1;
+  bool disallow_writes = 2;
+  bool disallow_lists = 3;
+  bool disallow_deletes = 4;
+
+  // if one of these two lists are not empty, restrict accesses
+  // to the given values
+  repeated bytes buckets = 10;
+  repeated bytes encrypted_path_prefixes = 11;
+
+  // if set, the validity time window
+  google.protobuf.Timestamp not_after = 20 [(gogoproto.stdtime) = true];
+  google.protobuf.Timestamp not_before = 21 [(gogoproto.stdtime) = true];
+}
+
+message Action {
+  enum Operation {
+    READ = 0;
+    WRITE = 1;
+    LIST = 2;
+    DELETE = 3;
+  }
+
+  Operation op = 1;
```

Comment:
Just for my understanding, the Operation isn't required, then if it isn't set, it defaults to `READ`, right?

Reply:
operation is required, but read seemed like a good default

Reply:
perhaps you're saying we should make an "INVALID" operation that is always rejected and make that the zero value?

Reply:
sorry, I rechecked protobuf docs and I realized that `required` doesn't apply for proto 3, I bore in mind the field rule from my previous experience with protobuf which was long time ago and I used the v2.

Before writing this response I've documented myself a bit more regarding protobuf v3, for hoping to make more sense on what I've written.

In protobuf v3 all the fields are optional, that means that the sender can send it or not; when they are not sent, the fields get the [default value according its type](https://developers.google.com/protocol-buffers/docs/proto3#default) and in enums is the first defined enum value.

The actions is required for the logic, so my preference is to explicitly define the operation rather than having a default one. I always see that's more straightforward to understand because the reader doesn't need to know the default value applied when it isn't specified. So, yes, I would prefer to have an _INVALID_ operation as default that will allow to know that the user actually forget to determine the operation and we can report an error about it.

Nonetheless, I agree that if we make a default valid operation, _READ_ is a good default.

Reply:
oh man, yeah, every single protobuf i used in v2 i marked all the fields optional cause protobuf v2 required was badness.

okay, i'll make an invalid type explicitly. i agree with you